portability-spi-cloud/src/main/java/org/dataportabilityproject/spi/cloud/storage/JobStore.java

```diff
@@ -55,12 +55,29 @@ default void update(UUID jobId, LegacyPortabilityJob job, JobAuthorization.State
   }
 
   /**
-   * Atomically updates the entry for {@code job}'s ID to {@code job}.
+   * Verifies a {@code PortabilityJob} already exists for {@code jobId}, and updates the entry to
+   * {@code job}.
    *
-   * @throws IOException if there was a problem updating the job.
+   * @throws IOException if a job didn't already exist for {@code jobId} or there was a problem
+   * updating it
    */
   void updateJob(UUID jobId, PortabilityJob job) throws IOException;
 
+  /**
+   * Verifies a {@code PortabilityJob} already exists for {@code jobId}, and updates the entry to
+   * {@code job}.
+   *
+   * <p>For auth-state-changing operations, the {@code previousState} may be validated as part of
+   * the transaction. Set {@code previousState} to null if no validation is desired.
+   *
+   * @throws IOException if a job didn't already exist for {@code jobId} or there was a problem
+   * updating it
+   * @throws IllegalStateException if state validation was requested ({@code previousState} was
+   * non-null) and the job's state was not the expected {@code previousState}
+   */
+  void updateJob(UUID jobId, PortabilityJob job, JobAuthorization.State previousState)
```

Comment:
I'm weary of adding the third param simply for validation since this class mostly just cares about storage. Since we usually have the existing job object before updateJob is called, can we validate before the updateJob method is called rather than having it leak into the JobStore?

Reply:
Good description of the issue. Ok, let's move forward with this.

It's up to you if you want to do this now or later, but maybe we can pass in an JobUpdateValidator interface that the client can provide to wrap this logic.

 void updateJob(UUID jobId, PortabilityJob job, JobUpdateValidator validator)

Something like
inteface JobUpdateValidator {
  /** Validates before update. */
  void validate(PortabilityJob previous, PortabilityJob updated);

Reply:
Nice, I like that solution, I went with that